### PR TITLE
Add ZIP 32 Arbitrary Key derivation

### PR DESCRIPTION
**Pull request by str4d**
Includes the FFI 0.11.0 build with its associated changes.

# Author

- [x] Self-review: Did you review your own code in GitHub's web interface? Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs.
- [x] Automated tests: Did you add appropriate automated tests for any code changes?
- [ ] Code coverage: Did you check the code coverage report for the automated tests?  While we are not looking for perfect coverage, the tool can point out potential cases that have been missed.
- [ ] Documentation: Did you update Docs as appropiate? (E.g [README.md](../blob/main/README.md), etc.)
- [ ] Run the app: Did you run the app and try the changes? 
- [ ] Did you provide Screenshots of what the App looks like before and after your changes as part of the description of this PR? (only applicable to UI Changes)
- [ ] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit.

# Reviewer

- [ ] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md) checklist?
- [ ] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?_You will find manual testing guidelines under our [manual testing section](../blob/mater/docs/testing/manual_testing)_
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage befor and after your changes and when possible, leave it in a better place. [Learn More...](../blob/master/docs/testing/local_coverage.md)_
- [ ] Documentation: Did you review Docs, [README.md](../blob/master/README.md), [LICENSE.md](../blob/master/LICENSE.md), and [Architecture.md](../blob/master/docs/Architecture.md) as appropriate?
- [ ] Run the app: Did you run the app and try the changes? While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data.

### CODE REVIEW

File: Sources/ZcashLightClientKit/Tool/DerivationTool.swift [hunk 1]
```diff
@@ -82,14 +109,46 @@ public class DerivationTool: KeyDeriving {
...
     }
+
+    /// Derives and returns a ZIP 32 Arbitrary Key from the given seed at the "wallet level", i.e.
+    /// directly from the seed with no ZIP 32 path applied.
+    ///
+    /// The resulting key will be the same across all networks (Zcash mainnet, Zcash testnet,
+    /// OtherCoin mainnet, and so on). You can think of it as a context-specific seed fingerprint
+    /// that can be used as (static) key material.
+    ///
+    /// - Parameter contextString: a globally-unique non-empty sequence of at most 252 bytes that identifies the desired context.
```
Comment (daira) [inline Sources/ZcashLightClientKit/Tool/DerivationTool.swift:128]: If failing to meet this requirement can cause a panic, it must also be documented at this layer.

File: Sources/ZcashLightClientKit/Tool/DerivationTool.swift [hunk 2]
```diff
@@ -48,6 +48,33 @@ public protocol KeyDeriving {
...
+    /// - Parameter accountNumber: `Int` with the account number
+    /// - Throws:
+    ///     - `derivationToolInvalidAccount` if the `accountIndex` is invalid.
+    ///     - some `ZcashError.rust*` error if the derivation fails.
+    /// - Returns a `[Uint8]`
+    static func deriveArbitraryWalletKey(contextString: [UInt8], seed: [UInt8]) throws -> [UInt8]
+
+    /// Derives and returns a ZIP 32 Arbitrary Key from the given seed at the account level.
+    ///
+    /// - Parameter contextString: a globally-unique non-empty sequence of at most 252 bytes that identifies the desired context.
```
Comment (daira) [inline Sources/ZcashLightClientKit/Tool/DerivationTool.swift:70]: If failing to meet this requirement can cause a panic, it must also be documented at this layer.

File: Sources/ZcashLightClientKit/Tool/DerivationTool.swift [hunk 3]
```diff
@@ -48,6 +48,33 @@ public protocol KeyDeriving {
...
     static func getAddressMetadata(_ addr: String) -> AddressMetadata?
+
+    /// Derives and returns a ZIP 32 Arbitrary Key from the given seed at the "wallet level", i.e.
+    /// directly from the seed with no ZIP 32 path applied.
+    ///
+    /// The resulting key will be the same across all networks (Zcash mainnet, Zcash testnet,
+    /// OtherCoin mainnet, and so on). You can think of it as a context-specific seed fingerprint
+    /// that can be used as (static) key material.
+    ///
+    /// - Parameter contextString: a globally-unique non-empty sequence of at most 252 bytes that identifies the desired context.
```
Comment (daira) [inline Sources/ZcashLightClientKit/Tool/DerivationTool.swift:59]: If failing to meet this requirement can cause a panic, it must also be documented at this layer.

File: Sources/ZcashLightClientKit/Tool/DerivationTool.swift [hunk 4]
```diff
@@ -82,14 +109,46 @@ public class DerivationTool: KeyDeriving {
...
+    ///     - `derivationToolInvalidAccount` if the `accountIndex` is invalid.
+    ///     - some `ZcashError.rust*` error if the derivation fails.
+    /// - Returns a `[Uint8]`
+    public static func deriveArbitraryWalletKey(contextString: [UInt8], seed: [UInt8]) throws -> [UInt8] {
+        return try ZcashKeyDerivationBackend.deriveArbitraryWalletKey(contextString: contextString, from: seed)
+    }
+
+    /// Derives and returns a ZIP 32 Arbitrary Key from the given seed at the account level.
+    ///
+    /// - Parameter contextString: a globally-unique non-empty sequence of at most 252 bytes that identifies the desired context.
```
Comment (daira) [inline Sources/ZcashLightClientKit/Tool/DerivationTool.swift:141]: If failing to meet this requirement can cause a panic, it must also be documented at this layer.

**Review by daira — APPROVED** [review]
utACK with blocking comments.

**Review by daira — APPROVED** [review]
utACK

**Review by nuttycom — APPROVED** [review]
utACK 1a924616903b9220777d4f016d283d5743037a32

**Pull request merged**